### ISSUE TITLE
Incompatible types in assignment with dynamic `__setattr__` and `__init_subclass__` for custom data types

### ISSUE THREAD

**Issue by ThorstenBrach** [microsoft/pyright#11204]
**Environment**
*   VS Code Version: 1.107.1
*   Pylance Version: v2025.10.4
*   Python Version: 3.13.3
*   OS: Windows

**Description**

Pylance's static type checker incorrectly reports an `Incompatible types in assignment` (reportAssignmentType) error when assigning a primitive Python type (e.g., `bool`, `int`) to a class attribute that is annotated with a custom `ctypes`-based class.

The issue occurs in a setup designed to mimic industrial PLC data types (IEC 61131-3), where a base class (`IEC_Struct`) dynamically wraps its annotated fields in a descriptor (`_FieldProxy`) that handles the type conversion at runtime. The code is valid and runs correctly, but Pylance fails to understand this dynamic behavior because it relies on static analysis.

**Code to Reproduce**

This is a minimal example that reproduces the issue. The `IEC_Struct` uses a `_FieldProxy` to intercept assignments to its annotated fields.

```python
import ctypes
from typing import Any

# --- Custom IEC-like Types ---
class BOOL(ctypes.c_bool):
    """Custom boolean type."""
    pass

class INT(ctypes.c_int16):
    """Custom integer type."""
    pass

# --- Field Proxy Descriptor ---
class _FieldProxy:
    """A descriptor that intercepts get/set operations on a ctypes storage."""
    def __init__(self, storage: ctypes.Structure, field_name: str):
        self._storage = storage
        self._field_name = field_name

    def __get__(self, obj, objtype=None):
        return getattr(self._storage, self._field_name)

    def __set__(self, obj, value: Any) -> None:
        # At runtime, this correctly handles the assignment.
        setattr(self._storage, self._field_name, value)

# --- Base Structure with Dynamic Wrappers ---
class IEC_Struct:
    """Base class that dynamically wraps annotated fields with a proxy."""
    def __init__(self):
        fields = []
        for name, type_hint in self.__annotations__.items():
            if issubclass(type_hint, ctypes._SimpleCData):
                fields.append((name, type_hint))
        
        StorageType = type(f"{self.__class__.__name__}_Storage", (ctypes.Structure,), {"_fields_": fields})
        self._storage = StorageType()

        for name, _ in self.__annotations__.items():
            proxy = _FieldProxy(self._storage, name)
            super().__setattr__(name, proxy)

    def __setattr__(self, name: str, value: Any):
        # Delegate assignment to the proxy if it exists
        if name in self.__dict__:
            super().__setattr__(name, value)
        else:
            super().__setattr__(name, value)


# --- User Implementation ---
class MyData(IEC_Struct):
    # User wants to declare clean, PLC-like types
    Flag: BOOL
    Counter: INT

# --- The Problem ---
instance = MyData()

# Pylance incorrectly flags these lines with:
# "Incompatible types in assignment (expression has type "bool", variable has type "BOOL")"
instance.Flag = True 

# "Incompatible types in assignment (expression has type "int", variable has type "INT")"
instance.Counter = 123

# The code runs without errors.
print(f"Flag value: {instance.Flag.value}, Counter value: {instance.Counter.value}")
```

**Expected Behavior**

Pylance should not report an error for the assignments `instance.Flag = True` and `instance.Counter = 123`. It should ideally recognize that the assignment is being handled by a custom `__setattr__` or a descriptor (`_FieldProxy`) on the `IEC_Struct` class, which makes the assignment valid at runtime.

**Actual Behavior**

Pylance reports a false-positive `reportAssignmentType` error, forcing users to either:
1.  Use `# type: ignore`, which is undesirable.
2.  Change the clean annotation `Flag: BOOL` to `Flag: Union[BOOL, bool]`, which clutters the code and defeats the purpose of creating clear, domain-specific types.
3.  Globally disable type-checking rules, which undermines the benefits of static analysis.

**Possible Solutions or Enhancements**

This issue highlights the conflict between Python's dynamic capabilities and static analysis. While full support for tracing dynamic `__setattr__` calls might be complex, a more pragmatic, intermediate solution would be to allow library authors to explicitly mark their types as "assignable from primitives". This would be a workaround, but a very effective one.

The proposal is to introduce a mechanism to disable type-checking for a specific type via a decorator or a special comment.

For example:

```python
# Proposal 1: Using a decorator
from pylance.lib import accepts_primitive_assignment

@accepts_primitive_assignment
class BOOL(ctypes.c_bool):
    ...

# Proposal 2: Using a special comment
class INT(ctypes.c_int16): # pylance: allow-primitive-assignment
    ...
```

When Pylance encounters an attribute annotated with a type marked in this way (e.g., `my_attr: BOOL`), it would automatically suppress `reportAssignmentType` errors for assignments of corresponding primitive types (`bool` for `BOOL`, `int` for `INT`, etc.) to that attribute.

This approach has several advantages:
*   **Granular Control:** It keeps type checking active for all other types in the codebase.
*   **Library-Side Solution:** The responsibility is on the library author, who knows the type is safe. Users of the library get a clean experience without needing `# type: ignore` or changing their annotations.
*   **Acknowledges Static Limitations:** It serves as a clear acknowledgement that this is a necessary workaround for a dynamic pattern that is difficult for static analysis to trace.

This would be a significant improvement over disabling rules globally or forcing users to clutter their code with `Union` or `type: ignore`.

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the Python typing spec. A value of type `builtins.bool` is not assignable to type `ctypes.BOOL`.

For details about type checking rules within the Python type system, refer to the [Python typing spec](https://typing.python.org/en/latest/spec/index.html).

If you have general questions about the type system, the [Python typing forum](https://discuss.python.org/c/typing/32) is a good place to post questions.
